Source/PINDiskCache.m

```diff
@@ -424,45 +433,61 @@ - (BOOL)_locked_createCacheDirectory
     
+    // while this may not be true if success is false, it's better than deadlocking later.
+    _diskWritable = YES;
+    
     return success;
 }
 
-- (void)_locked_initializeDiskProperties
+- (void)initializeDiskProperties
 {
     NSUInteger byteCount = 0;
     NSArray *keys = @[ NSURLContentModificationDateKey, NSURLTotalFileAllocatedSizeKey ];
     
     NSError *error = nil;
-    NSArray *files = [[NSFileManager defaultManager] contentsOfDirectoryAtURL:_cacheURL
-                                                   includingPropertiesForKeys:keys
-                                                                      options:NSDirectoryEnumerationSkipsHiddenFiles
-                                                                        error:&error];
+    
+    [self lock];
+        NSArray *files = [[NSFileManager defaultManager] contentsOfDirectoryAtURL:_cacheURL
+                                                       includingPropertiesForKeys:keys
+                                                                          options:NSDirectoryEnumerationSkipsHiddenFiles
+                                                                            error:&error];
+    [self unlock];
+    
     PINDiskCacheError(error);
     
     for (NSURL *fileURL in files) {
         NSString *key = [self keyForEncodedFileURL:fileURL];
         
         error = nil;
-        NSDictionary *dictionary = [fileURL resourceValuesForKeys:keys error:&error];
-        PINDiskCacheError(error);
         
-        NSDate *date = [dictionary objectForKey:NSURLContentModificationDateKey];
-        if (date && key)
-            [_dates setObject:date forKey:key];
+        // Do not continue to hold the lock while processing files.
+        [self lock];
+            NSDictionary *dictionary = [fileURL resourceValuesForKeys:keys error:&error];
+            PINDiskCacheError(error);
         
-        NSNumber *fileSize = [dictionary objectForKey:NSURLTotalFileAllocatedSizeKey];
-        if (fileSize) {
-            [_sizes setObject:fileSize forKey:key];
-            byteCount += [fileSize unsignedIntegerValue];
-        }
+            NSDate *date = [dictionary objectForKey:NSURLContentModificationDateKey];
+            if (date && key)
+                [_dates setObject:date forKey:key];
         
-        [_knownKeys addObject:key];
+            NSNumber *fileSize = [dictionary objectForKey:NSURLTotalFileAllocatedSizeKey];
+            if (fileSize) {
+                [_sizes setObject:fileSize forKey:key];
+                byteCount += [fileSize unsignedIntegerValue];
+            }
+        
+            [_knownKeys addObject:key];
+        [self unlock];
     }
     
-    if (byteCount > 0)
-        _byteCount = byteCount;
+    [self lock];
+        if (byteCount > 0)
+            _byteCount = byteCount;
+    
+        if (self->_byteLimit > 0 && self->_byteCount > self->_byteLimit)
+            [self trimToSizeByDateAsync:self->_byteLimit completion:nil];
```

Comment:
It looks like we may need a way to avoid scheduling a large number of trim operations while this is processing. It probably makes sense to have a _trimPending / scheduled variable, so that a series of setObject: calls during this time doesn't schedule a lot of them.

Since the operation queue is concurrent, it looks like there may be an issue with multiple trim operations starting in parallel too — this could be addressed by using a barrier block on the underlying queue, which GCD will ensure is run serially.

Reply:
@appleguy the tim's are actually coalesced thanks to @nguyenhuy's improvement in PINOperationQueue! We could also skip trimming until we have a known state.